src/share/poudriere/common.sh

```diff
@@ -3105,16 +3126,61 @@ prefix_output() {
 	prefix_stderr "${extra}" prefix_stdout "${extra}" "$@"
 }
 
+filter_raw_port_deps() {
+	tr ' ' '\n' | \
+	sed -E "s,(((${GEN_PREFIXES})[0-9]+)-.*)/(${GEN_PREFIXES})-(.*),\1/\2-\5,g" | \
+	sed -e "s,[[:graph:]]*/usr/ports/,,g" \
+	-e "s,:[[:graph:]]*,,g" -e '/^$/d' | \
+	sort -u
+}
+
+filter_raw_pkg_deps() {
+	sed -E "s,(((${GEN_PREFIXES})[0-9]+)-.*)/(${GEN_PREFIXES})-(.*),\1/\2-\5,g" | \
+	sed -e "s,.*:,,g"
+}
+
 list_deps() {
-	[ $# -ne 1 ] && eargs list_deps directory
+	[ $# -ne 2 ] && eargs list_deps directory gen
 	local dir="/usr/ports/$1"
 	local makeargs="-VPKG_DEPENDS -VBUILD_DEPENDS -VEXTRACT_DEPENDS -VLIB_DEPENDS -VPATCH_DEPENDS -VFETCH_DEPENDS -VRUN_DEPENDS"
 
 	prefix_stderr_quick "(${COLOR_PORT}$1${COLOR_RESET})${COLOR_WARN}" \
-		injail make -C ${dir} $makeargs | \
-		sed -e "s,[[:graph:]]*/usr/ports/,,g" \
-		-e "s,:[[:graph:]]*,,g" -e '/^$/d' | tr ' ' '\n' | \
-		sort -u || err 1 "Makefile broken: $1"
+		injail_port_make ${1} "${2}" "$makeargs" | \
+		filter_raw_port_deps || err 1 "Makefile broken: $1"
+}
+
+genport_extract_port() {
+	echo "$1" | sed -E "s,/((${GEN_PREFIXES})[0-9]+)-,/\2-,g"
+}
+
+genport_extract_gen() {
+	local str=$(echo "$1" | sed -E "s,^.*/(${GEN_PREFIXES})([0-9]*)-.*$,\2,g")
+	[ "$str" != "$1" ] && echo "$str"
+        return 0
+}
+
+genport_extract_prefix() {
+	local str=$(echo "$1" | sed -E "s,^(.*/)(${GEN_PREFIXES})([0-9]*)-.*$,\2,g")
+	[ "$str" != "$1" ] && echo "$str"
+        return 0
+}
+
+genport_construct() {
+	local port="$1"
+	local gen="$2"
+	echo "$1" | sed -E "s,^(.*/)(${GEN_PREFIXES})-(.*)$,\1\2${gen}-\3,g"
+}
+
+make_env_gen() {
+	local port="$1"
+	local gen="$2"
+	[ -z "${gen}" ] && return
+	case $(genport_extract_prefix $1) in
+		py) echo "PYTHON_VERSION=$(echo $gen | sed -E "s/(.)(.)/\1.\2/g")"
+		;;
+		ruby) echo "RUBY_VERSION=$gen"
```

Comment:
This is fundamentally wrong. Poudriere should not have such specific handling here.